Biome API: Implement biome-defined air nodes

paramat opened the pull request:
WIP in long-term development and testing.
For #5725 
If this is approved i will create new PRs rebased on master, as rebasing all this will be not worth it.
This is not for 5.0.0, too big and potentially problematic for such a near release.

* Nodedef:
Resurrect 'air_equivalent = bool' nodedef actually stated for "air" but not used anywhere. This is used for nodes that should be treated by the engine like "air" is now. Using "airlike" drawtype for this purpose is not ideal as visible nodes are also wanted.
* Record the new field in network protocol.

////////////////
* General mapgen:
Add "mapgen_air" alias as the 4th base terrain node, falls back to "air" to not break games.
* Mapgens:
Place the new base terrain node "mapgen_air" instead of hardcoded "air".
Avoid overgeneration removing biome-defined air.
* Biomes:
Add "node_air" field as the new biome-defined air node.
* Biome generation:
Detect this new base terrain node and place biome-defined air nodes.
Update dust placement.
* Decorations:
Allow simple decorations to be placed in 'air_equivalent' without using forced placement.
* Schematic decorations:
Allow schematic nodes to be placed in 'air_equivalent' without using forced placement.
* Cavegen:
For tunnels and caverns: Make generation similar to biome generation. calculating biome at node resolution and placing the biome-defined air nodes when excavating.
For large randomwalk caves: The biome for each cave segment is already calculated for biome-defined cave liquids, for each segment place the biome-defined air node when excavating.
* Dungeons:
The biome for each dungeon is already calculated for biome-defined dungeon nodes, place the biome-defined air node when excavating the rooms and corridors.

///////////////////
* Player spawn:
Allow player to spawn in 'air_equivalent' nodes. 
* Liquid processing:
When a flowing liquid disappears from a node due to source removal replace with any detected 'air_equivalent' face-connected neighbour node. If none fallback to "air".
* Dug nodes and minetest.remove_node():
Replace with any detected 'air_equivalent' face-connected neighbour node. If none fallback to "air".

////////////////////
* Add a planetoid core mapgen for testing biome air / space mapgen issues. Not for merging.

paramat commented:
Current bug, pasting here for my use:
```Thread 15 "Emerge-0" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffbbfff700 (LWP 12347)]
0x0000000000a420ae in MapgenBasic::generateBiomes (this=0x24f1a70)
    at /home/matstandard/minetestf/gitmods/minetest/src/mapgen/mapgen.cpp:753
753					vm->m_data[vi] = MapNode(biome->c_air);
(gdb) 
(gdb) bt full
#0  0x0000000000a420ae in MapgenBasic::generateBiomes (this=0x24f1a70)
    at /home/matstandard/minetestf/gitmods/minetest/src/mapgen/mapgen.cpp:753
        c = 126
        is_stone_surface = false
        is_water_surface = false
        y = 47
        biome_y_min = -31000
        vi = 211360
        air_above = true
        river_water_above = false
        water_above = false
        nplaced = 0
        base_filler = 0
        depth_water_top = 0
        depth_riverbed = 0
        c_above = 126
        biome = 0x0
        depth_top = 0
        x = 128
        z = -32
        __PRETTY_FUNCTION__ = "virtual void MapgenBasic::generateBiomes()"
        em = @0x7fffb457b6d4: {X = 112, Y = 112, Z = 112}
        index = 0
```

paramat commented:
https://github.com/minetest/minetest/blob/e98fd934ce17a7af0b4629fc88158373303c88a0/builtin/game/register.lua#L352

https://github.com/minetest/minetest/blob/e98fd934ce17a7af0b4629fc88158373303c88a0/builtin/game/register.lua#L368

"air" and "ignore" have 'air_equivalent = true,' but i don't think it's used anywhere else, maybe we can use it now.

CoderForTheBetter commented:
Although not completely related to development, just wanted to say that I'm very excited about this and will try to help test this in the future.

Ezhh commented:
When you dig a node in a biome that has an alternative air defined, will that take the place of the dug node?

paramat commented:
Ezhh that is one example of why this will be difficult work, the engine is based on the assumption that empty generated world is the hardcoded node "air".
In my 'moonrealm' mod Lua is used to search for a nearby vacuum or 'life-support air' node to add when a node is dug. The same could perhaps be done in the engine.
I fully expect this PR feature to be imperfect due to the issues but i think it's worth doing anyway and slowly addressing the issues over time. Lua mods can provide the missing functionality on the meantime.
My 'moonrealm' mod works quite well but mapgen is slow because i can't place vacuum nodes using the engine, this PR feature will make space games and mods doable.

paramat commented:
Now adds the nodedef 'air_equivalent = true' that is already stated for "air" but unused.

paramat commented:
![screenshot_20180415_222300](https://user-images.githubusercontent.com/3686677/38783579-d4a37f5a-40fb-11e8-9873-79b7f42b1a18.png)

![screenshot_20180415_222426](https://user-images.githubusercontent.com/3686677/38783581-d83e19ea-40fb-11e8-94cf-086f4eb086be.png)

TMcSquared commented:
Woah! that would be awesome!

paramat commented:
Implemented for all cavegen, dungeons next.

TMcSquared commented:
This could be used to make a sort of mood for each biome, i.e. the sky is a different tone or something like that. Or..wait..wouldn't that be a different PR entirely?

paramat commented:
I don't want to get people's hopes up for this, it's still very buggy. I am seeing that making this work will make cave generation more complex and a little slower. It's interesting to try but i'm still not sure whether it will be merged, it's somewhat of a nightmare and it may be too much trouble. However i will do my best :) If i get it working it can be profilied for effect on mapgen time.

TMcSquared commented:
Ah understood :grin:

paramat commented:
List of stuff done:

* Nodedef:
Resurrect 'air_equivalent = bool' nodedef actually stated for "air" but not used anywhere.
* Record the new field in network protocol.
* General mapgen:
Add "mapgen_air" alias as the 4th base terrain node, falls back to "air" to not break games.
* Mapgens:
Place the new base terrain node "mapgen_air" instead of hardcoded "air".
Avoid overgeneration removing biome-defined air.
* Biomes:
Add "node_air" field as the new biome-defined air node.
* Biome generation:
Detect this new base terrain node and place biome-defined air nodes.
Update dust placement.
* Decorations:
Allow simple decorations to be placed in 'air_equivalent' without using forced placement.
* Schematic decorations:
Allow schematic nodes to be placed in 'air_equivalent' without using forced placement.
* Cavegen:
For tunnels and caverns: Make generation similar to biome generation. calculating biome at node resolution and placing the biome-defined air nodes when excavating.
For large randomwalk caves: The biome for each cave segment is already calculated for biome-defined cave liquids, for each segment place the biome-defined air node when excavating.
* Dungeons:
The biome for each dungeon is already calculated for biome-defined dungeon nodes, place the biome-defined air node when excavating the rooms and corridors.
* Player spawn:
Allow player to spawn in 'air_equivalent' nodes.

paramat commented:
Now all the mapgen stuff is done, but coping with 'air_equivalent' nodes after mapgen is not yet done, for example placing or spreading biome air instead of "air" when a node is dug, or water flows away etc. However these things can already be done in Lua by a mod, at least mapgen will not be slowed by using a luavoxelmanip to place biome air. So this work is already useful as it is, even if further integration of 'air_equivalent' into the engine is problematic.

Next i need to profile mapgen to see the effect of the slightly more complex code, and do in-game testing to assess further issues.

paramat commented:
![screenshot_20180430_185553](https://user-images.githubusercontent.com/3686677/39444066-5b602bca-4cae-11e8-8084-80119df51ccb.png)

It's a 200x100x200 biome using the XYZ biome limits and with the biome air being a 'glasslike' cloud node.

paramat commented:
During testing i found the biggest continuously-connected large cave system i have ever seen:

![screenshot_20180426_175046](https://user-images.githubusercontent.com/3686677/39444966-5550f4f0-4cb1-11e8-8c76-3467fff1d7ef.png)

paramat commented:
Added a simple planetoid core mapgen for testing (not for merge).

![screenshot_20180501_195901](https://user-images.githubusercontent.com/3686677/39489312-f1319a84-4d7c-11e8-8ae5-acc58a0ed363.png)

^ Spawned, turned around and saw this, scariest spawn ever

![screenshot_20180501_200901](https://user-images.githubusercontent.com/3686677/39489355-1a44e58e-4d7d-11e8-995a-93f8942b915f.png)

^ Standing right in the corner of the planetoid, it has a misty 'glasslike drawtype' node as atmosphere

![screenshot_20180501_200939](https://user-images.githubusercontent.com/3686677/39489386-38882a2e-4d7d-11e8-8a5c-b5ddbc0ce480.png)

^ Outside the planetoid is vacuum nodes, note suffocation. The dot above the water is what i was standing on in the previous screenshot

The mod code to do this is:
```
minetest.clear_registered_biomes()
minetest.clear_registered_ores()
minetest.clear_registered_decorations()

minetest.register_biome({
	name = "land",
	node_top = "default:dirt_with_grass",
	depth_top = 1,
	node_filler = "default:dirt",
	depth_filler = 2,
	node_air = "testfogcloud:fog",
	max_pos = {x = 207, y = 207, z = 207},
	min_pos = {x = -192, y = 5, z = -192},
	heat_point = 50,
	humidity_point = 50,
})

minetest.register_biome({
	name = "lake",
	node_top = "default:sand",
	depth_top = 1,
	node_filler = "default:sand",
	depth_filler = 2,
	node_air = "testfogcloud:fog",
	max_pos = {x = 207, y = 4, z = 207},
	min_pos = {x = -192, y = -192, z = -192},
	heat_point = 50,
	humidity_point = 50,
})

minetest.register_biome({
	name = "space",
	node_air = "testfogcloud:vacuum",
	max_pos = {x = 31000, y = 31000, z = 31000},
	min_pos = {x = -31000, y = -31000, z = -31000},
	heat_point = 1000,
	humidity_point = 1000,
})

minetest.register_node("testfogcloud:fog", {
	description = "Fog",
	drawtype = "glasslike",
	tiles = {"testfogcloud_fog.png"},
	paramtype = "light",
	is_ground_content = false,
	sunlight_propagates = true,
	walkable = false,
	pointable = false,
	diggable = false,
	buildable_to = true,
	floodable = true,
	air_equivalent = true,
	post_effect_color = {a = 31, r = 241, g = 248, b = 255},
})

minetest.register_node("testfogcloud:vacuum", {
	description = "Vacuum",
	drawtype = "airlike",
	paramtype = "light",
	is_ground_content = false,
	sunlight_propagates = true,
	walkable = false,
	pointable = false,
	diggable = false,
	buildable_to = true,
	floodable = false,
	air_equivalent = true,
	drowning = 1,
})
```
Note how the 'space' biome uses silly heat and humidity points to make sure this biome never occurs in a planetoid:
```
	heat_point = 1000,
	humidity_point = 1000,
```
Because each planetoid has biomes defined with XYZ limits matching the planetoid cube, outside these limits only the space biome is active, placing vacuum.

CoderForTheBetter commented:
@paramat This looks great. I hope it turns out to not impact map generation too much, I really want to see this in 0.5.0!

TMcSquared commented:
hmm regarding an example screenshot you showed @paramat , would this make it possible to create actual 3d clouds(not just flat blocky ones)?

paramat commented:
> would this make it possible to create actual 3d clouds

No because you'd need one biome per cloud cuboid, but you can create a cloud biome that is one big cuboid, as shown in https://github.com/minetest/minetest/pull/7241#issuecomment-385490738

paramat commented:
I realised that the liquid processing code already searches through neighbour nodes, so it is trivial to make any neighbouring 'air_equivalent' node the one placed if liquid flows away from a node. This simulates a neighbouring 'air_equivalent' node flowing into the space instead of "air" always being placed there.

![screenshot_20180502_181932](https://user-images.githubusercontent.com/3686677/39539136-cadccb78-4e36-11e8-9bce-b885f0711dd5.png)

^ On my test planetoid with atmosphere made visible as a 'glasslike' drawtype node, liquid placed.

![screenshot_20180502_181953](https://user-images.githubusercontent.com/3686677/39539202-ef1a5f00-4e36-11e8-9444-b91eee80b463.png)
 ^ Remove the source. Once liquid flows away no bubbles of "air" are present.

TMcSquared commented:
@Ferk I meant 3d as in not just a single layer of cubes

Ferk commented:
@TMcSquared  the default thickness of 3dclouds is 16 layers of nodes, though it's customizable with modding:
https://github.com/minetest/minetest/blob/master/doc/lua_api.txt#L4208

TMcSquared commented:
it still does not have any roundedness on the vertical sides

Dumbeldor commented on src/mapgen/cavegen.cpp:
```diff
@@ -240,6 +251,8 @@ bool CavernsNoise::generateCaverns(MMVManip *vm, v3s16 nmin, v3s16 nmax)
 		// Initial 3D noise index at column top
 		u32 index3d = (z - nmin.Z) * m_zstride_1d + m_csize.Y * m_ystride +
 			(x - nmin.X);
+		Biome *biome = NULL;
```
nullptr

paramat replied:
Thanks will do.

Dumbeldor commented on src/mapgen/cavegen.cpp:
```diff
@@ -89,11 +87,12 @@ void CavesNoiseIntersection::generateCaves(MMVManip *vm,
-		u16 base_filler = depth_top + biome->depth_filler;
-		u16 depth_riverbed = biome->depth_riverbed;
+
+		Biome *biome = NULL;
```
nullptr

Dumbeldor commented on src/mapgen/cavegen.h:
```diff
@@ -29,26 +29,20 @@ class GenerateNotifier;
 private:
 	const NodeDefManager *m_ndef;
-	BiomeManager *m_bmgr;
+	BiomeGen *m_bmgn;
```
BiomeGen *m_bmgn = nullptr

paramat replied:
I'm not sure it needs to be set to nullptr here and in cavernsNoise, as BiomeGen is required and not optional as in cavesRandomWalk.
However i will use it here:
```

 	CavesRandomWalk(const NodeDefManager *ndef, GenerateNotifier *gennotify =
 		NULL, s32 seed = 0, int water_level = 1, content_t water_source =
 		CONTENT_IGNORE, content_t lava_source = CONTENT_IGNORE,
 		int lava_depth = -256, BiomeGen *biomegen = NULL);
```

Dumbeldor commented on src/mapgen/cavegen.h:
```diff
@@ -69,14 +63,15 @@ class CavernsNoise
 
 private:
 	const NodeDefManager *m_ndef;
+	BiomeGen *m_bmgn;
```
BiomeGen *m_bmgn = nullptr;

paramat commented:
Since this is long term experimentation, closing for now.
